src/swap.rs

```diff
@@ -0,0 +1,138 @@
+use rustc::lint::*;
+use rustc_front::hir::*;
+use syntax::codemap::mk_sp;
+use utils::{differing_macro_contexts, snippet_opt, span_lint_and_then, SpanlessEq};
+
+/// **What it does:** This lints manual swapping.
+///
+/// **Why is this bad?** The `std::mem::swap` function exposes the intent better without
+/// deinitializing or copying either variable.
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust,ignore
+/// let t = b;
+/// b = a;
+/// a = t;
+/// ```
+declare_lint! {
+    pub MANUAL_SWAP,
+    Warn,
+    "manual swap"
+}
+
+/// **What it does:** This lints `foo = bar; bar = foo` sequences.
+///
+/// **Why is this bad?** This looks like a failed attempt to swap.
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust,ignore
+/// a = b;
+/// b = a;
+/// ```
+declare_lint! {
+    pub SUSPICIOUS_SWAP,
```

Comment:
I think we should call this `SWAP_FAIL` or `ALMOST_SWAPPED`. The way it is it's not a swap.